ext/cache/sqlite.rs

```diff
@@ -0,0 +1,401 @@
+      db.execute(
+        "INSERT OR IGNORE INTO cache_storage (cache_name) VALUES (?1)",
+        params![cache_name],
+      )?;
+      let cache_id = db.query_row(
+        "SELECT id FROM cache_storage WHERE cache_name = ?1",
+        params![cache_name],
+        |row| {
+          let id: i64 = row.get(0)?;
+          Ok(id)
+        },
+      )?;
+      Ok::<i64, AnyError>(cache_id)
+    })
+    .await?
+  }
+
+  /// Check if a cache with the provided name exists.
+  /// Note: this doesn't check the disk, it only checks the sqlite db.
+  async fn storage_has(&self, cache_name: String) -> Result<bool, AnyError> {
+    let db = self.connection.clone();
+    tokio::task::spawn_blocking(move || {
+      let db = db.lock();
+      let cache_exists = db.query_row(
+        "SELECT count(cache_name) FROM cache_storage WHERE cache_name = ?1",
+        params![cache_name],
+        |row| {
+          let count: i64 = row.get(0)?;
+          Ok(count > 0)
+        },
+      )?;
+      Ok::<bool, AnyError>(cache_exists)
+    })
+    .await?
+  }
+
+  /// Delete a cache storage. Internally, this deletes the row in the sqlite db.
+  async fn storage_delete(&self, cache_name: String) -> Result<bool, AnyError> {
+    let db = self.connection.clone();
+    let cache_storage_dir = self.cache_storage_dir.clone();
+    tokio::task::spawn_blocking(move || {
+      let db = db.lock();
+      let maybe_cache_id = db
+        .query_row(
+          "DELETE FROM cache_storage WHERE cache_name = ?1 RETURNING id",
+          params![cache_name],
+          |row| {
+            let id: i64 = row.get(0)?;
+            Ok(id)
+          },
+        )
+        .optional()?;
+      if let Some(cache_id) = maybe_cache_id {
+        let cache_dir = cache_storage_dir.join(cache_id.to_string());
+        if cache_dir.exists() {
+          std::fs::remove_dir_all(cache_dir)?;
+        }
+      }
+      Ok::<bool, AnyError>(maybe_cache_id.is_some())
+    })
+    .await?
+  }
+
+  async fn put(
+    &self,
+    request_response: CachePutRequest,
+  ) -> Result<Option<Rc<dyn Resource>>, AnyError> {
+    let db = self.connection.clone();
+    let cache_storage_dir = self.cache_storage_dir.clone();
+    let maybe_body_path = tokio::task::spawn_blocking(move || {
+      let response_body_key = if request_response.response_has_body {
+        Some(hash(&request_response.request_url))
+      } else {
+        None
+      };
+      let maybe_response_body = {
+        let db = db.lock();
+        db.query_row(
+          "INSERT OR REPLACE INTO request_response_list
+               (cache_id, request_url, request_headers, response_headers, response_body_key, response_status, response_status_text)
+               VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
+               RETURNING response_body_key",
+          (
+            request_response.cache_id,
+            &request_response.request_url,
+            serde_json::to_string(&request_response.request_headers)?,
```

Comment:
JSON serialization/deserialization is relatively slow. Also we can't use `ByteString` with `serde_json`. We'll need some other encoding for `Vec<(ByteString, ByteString)>` -> `Vec<u8>`.

Reply:
What can be used here?

Reply:
We could do a simple custom serialisation in the form of `[ len name (u32) | name bytes | len val (u32) | val bytes ]*`. If there is something off the shelf we can use, I'd prefer that.

Reply:
Serialize into a buffer with V8 serializer? I'm not sure how stable that will be.

Reply:
That is stable.

Reply:
Since this isn't a hard blocker, I would like to add a TODO and do the optimization later. Any objections?